test_runner/batch_others/test_wal_acceptor_async.py

```diff
@@ -0,0 +1,166 @@
+            assert await cur.fetchone() == (self.n_accounts * self.init_amount, )
+
+    async def transfer(self, from_uid, to_uid, amount):
+        # avoid deadlocks by sorting uids
+        if from_uid > to_uid:
+            from_uid, to_uid, amount = to_uid, from_uid, -amount
+
+        async with self.conn.cursor() as cur:
+            await cur.execute('BEGIN')
+            await cur.execute('UPDATE bank_accs SET amount = amount + (%s) WHERE uid = %s', (
+                amount,
+                to_uid,
+            ))
+            assert cur.rowcount == 1
+            await cur.execute('UPDATE bank_accs SET amount = amount - (%s) WHERE uid = %s', (
+                amount,
+                from_uid,
+            ))
+            assert cur.rowcount == 1
+            await cur.execute('INSERT INTO bank_log VALUES (%s, %s, %s)', (
+                from_uid,
+                to_uid,
+                amount,
+            ))
+            await cur.execute('COMMIT')
+
+
+class WorkerStats(object):
+    def __init__(self, n_workers):
+        self.n_workers = n_workers
+        self.counters = [0] * n_workers
+        self.running = True
+
+    def reset(self):
+        self.counters = [0] * self.n_workers
+
+    def inc_progress(self, worker_id):
+        self.counters[worker_id] += 1
+
+    def check_progress(self):
+        # every worker should finish at least one tx
+        for i in range(self.n_workers):
+            assert self.counters[i] > 0
+
+        progress = sum(self.counters)
+        print('All workers made {} transactions'.format(progress))
+
+
+async def run_random_worker(stats: WorkerStats,
+                            pg: Postgres,
+                            worker_id,
+                            n_accounts,
+                            max_transfer,
+                            debug=False):
+    pg_conn = await pg.connect_async()
+    bank = BankClient(pg_conn)
+
+    if debug:
+        print('Started worker {}'.format(worker_id))
+
+    while stats.running:
+        from_uid = random.randint(0, n_accounts - 1)
+        to_uid = (from_uid + random.randint(1, n_accounts - 1)) % n_accounts
+        amount = random.randint(1, max_transfer)
+
+        await bank.transfer(from_uid, to_uid, amount)
+        stats.inc_progress(worker_id)
+
+        if debug:
+            print('Executed transfer({}) {} => {}'.format(amount, from_uid, to_uid))
+
+    if debug:
+        print('Finished worker {}'.format(worker_id))
+
+    await pg_conn.close()
+
+
+# This test will run several iterations and check progress in each of them.
+# On each iteration 1 acceptor is stopped, and 2 others should allow
+# background workers execute transactions. In the end, state should remain
+# consistent.
+async def run_restarts_under_load(pg: Postgres, acceptors: List[WalAcceptor], n_workers=10):
+    n_accounts = 100
+    init_amount = 100000
+    max_transfer = 100
+    period_time = 10
+    iterations = 6
+
+    pg_conn = await pg.connect_async()
```

Comment:
Would `async with` work here?

Reply:
Should work, but I'm not sure if it will be better